### DISCUSSION TITLE
can you use spawn in framework?

### DISCUSSION THREAD

**Discussion by gganapat** [observablehq/framework#1288] [Q&A]
Hi:

I tried to see if Observable Framework will allow using spawn, half expecting it to not work. Sure enough, I got error "TypeError: error loading dynamically imported module: node:child_process". BTW this code works if I do node xxx.js from terminal.

I would like to understand what is going on and if it is at all possible to get it to work. 

Thanks!
Gani --
```
import { spawn } from 'node:child_process';

function runWolframScript(input) {
    return new Promise((resolve, reject) => {
        const wolframScript = spawn('wolframscript', ['-code', input]);

        let output = '';
        let error = '';

        wolframScript.stdout.on('data', (data) => {
            output += data;
        });

        wolframScript.stderr.on('data', (data) => {
            error += data;
        });

        wolframScript.on('close', (code) => {
            if (code === 0) {
                resolve(output);
            } else {
                reject(error);
            }
        });
    });
}

const input = 'Log[20.,10]';
runWolframScript(input)
    .then((result) => {
        console.log('Result:', result);
    })
    .catch((error) => {
        console.error('Error:', error);
    });
```

**Comment by Fil** [comment]
It should work. I don't have wolfram on my machine, but the following data loader does display the list of files:

```
❯ cat docs/test-spawn.txt.js
import {spawn} from "node:child_process";

const ls = spawn("ls", ["-lh", "/usr"]);

ls.stdout.on("data", (data) => {
  console.log(`stdout: ${data}`);
});

ls.stderr.on("data", (data) => {
  console.error(`stderr: ${data}`);
});

ls.on("close", (code) => {
  console.log(`child process exited with code ${code}`);
});

```

**Comment by mbostock** [comment]
Yes, any Node program can work as a data loader. In a case like this you could also just write a shell script, or even register wolframscript as an interpreter.

**Reply by gganapat** [reply]
Thanks! In the past I have written shell scripts for wolframscript as data loader which have worked fine. I was just seeing if it can be done directly without using data loader. I guess that was my problem -- I was trying to do this directly in index.md as js script which didn't work.

Thanks again,
Gani --

**Reply by mbostock** [reply]
No, JavaScript in Markdown (and JavaScript you import into Markdown) runs in the browser, so you can’t use Node.js APIs there.

**Comment by gganapat** [comment]
Ok, thanks for the clarification.Gani-On Apr 27, 2024, at 10:57 AM, Mike Bostock ***@***.***> wrote:﻿
No, JavaScript in Markdown (and JavaScript you import into Markdown) runs in the browser, so you can’t use Node.js APIs there.
